### PR TITLE
Add additional padding layer to deconv converter to fix output_shape

### PR DESCRIPTION
**Pull request by pooyadavoodi**
- TensorFlow has the argument output_shape in Deconv API (conv2d_transpose) that specifies the output size, otherwise the output size is ambiguous if strides>1.
- TRT API for IDeconvolution doesn't have such an argument, and thus the size of the TRT output tensor is off if stride>1.
- This PR adds a padding layer on the output of Deconv if needed, in order to fix the output size.

### CODE REVIEW

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 1]
```diff
@@ -2229,8 +2229,25 @@ Status ConvertConv2DHelper(OpConverterParams* params, int group,
...
-
+  nvinfer1::ITensor* conv_output_tensor = conv_layer->getOutput(0);
+  nvinfer1::ITensor* output_tensor;
+  // Add an extra padding for Deconv because TRT doesn't accept the
+  // argument output_shape and thus the TRT output shape could be wrong
+  // in case of strides>1.
+  if (is_conv2d_backprop_input) {
+    auto tf_output_shape = backprop_output_size.GetTrtDims();
+    nvinfer1::Dims trt_output_shape = conv_output_tensor->getDimensions();
+    const int heightDiff = tf_output_shape.d[h_index - 1] - trt_output_shape.d[1];
```
Comment (aaroey) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: height_diff
width_diff below

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 2]
```diff
@@ -2229,8 +2229,25 @@ Status ConvertConv2DHelper(OpConverterParams* params, int group,
...
+    const int widthDiff = tf_output_shape.d[w_index - 1] - trt_output_shape.d[2];
+    nvinfer1::DimsHW pre_padding(0, 0);
+    nvinfer1::DimsHW post_padding(heightDiff, widthDiff);
```
Comment (aaroey) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: What happen if height/width_diff is 0? Please comment.
Also, is it possible to get negatives?

**Review by aaroey — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc (hunk 1 above)
Comment (pooyadavoodi) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: Done

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc (hunk 2 above)
Comment (pooyadavoodi) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: Done

**Review by aaroey — APPROVED** [review]

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 3]
```diff
@@ -2230,7 +2230,33 @@ Status ConvertConv2DHelper(OpConverterParams* params, int group,
     conv_layer = layer;
   }
   nvinfer1::ITensor* output_tensor = conv_layer->getOutput(0);
-
+  // Add an extra padding for Deconv because TRT doesn't accept the
+  // argument output_shape and thus the TRT output shape could be wrong
+  // in case of strides>1.
+  if (is_conv2d_backprop_input) {
+    auto tf_output_shape = backprop_output_size.GetTrtDims();
```
Comment (aaroey) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: I think this should be `backprop_output_size.weights().GetValues()`, but I'm fixing it. Just FYI.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 4]
```diff
@@ -2230,7 +2230,33 @@ Status ConvertConv2DHelper(OpConverterParams* params, int group,
...
+    nvinfer1::Dims trt_output_shape = output_tensor->getDimensions();
+    // What determines the padding size is the difference between the given
+    // input_sizes (tf_output_shape) and TRT computed size.
+    const int height_diff = tf_output_shape.d[h_index - 1] - trt_output_shape.d[1];
```
Comment (aaroey) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: Accordingly this should be tf_output_shape[h_index]? Similar below.

Comment (aaroey) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: Do we need to consider NHWC vs NCHW to determine whether to use trt_output_shape.d[1] or d[2]? Same below.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc (hunk 3 above)
Comment (pooyadavoodi) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: hmmm, I am wondering how the test worked?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc (hunk 4 above)
Comment (pooyadavoodi) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: Since `tf_output_shape = backprop_output_size.GetTrtDims()`, `tf_output_shape` is in TRT format which means it doesn't have  the batch dimension. That's why need `h_index - 1`.

Comment (aaroey) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: It seems the output is always NCHW? Then should it be d[2] instead?

Comment (pooyadavoodi) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: `trt_output_shape` is the output of TRT which is always NCHW.
